File: Beginner/C/Array/PlusOne.c
/**
 * Return an array of size *returnSize.
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* plusOne(int* digits, int digitsSize, int* returnSize) {
    int carry = 1;
    *returnSize = digitsSize;
    for (int i = digitsSize - 1; i >= 0; i --) {
        if (carry == 1) {
            if (digits[i] == 9)
                digits[i] = 0;
            else {
                digits[i] ++;
                carry = 0;
            }
        }
        else
            break;
    }
    int pos = 0;
    int* res = malloc(sizeof(int) * (digitsSize + 1));
    if (carry == 1) {
        (*returnSize) ++;
        res[0] = 1;
        pos ++;
    }
    for (int i = 0; pos < (*returnSize); pos ++, i ++) {
        res[pos] = digits[i];
    }
    return res;
}

File: Beginner/Java/Array/PlusOne.java
class Solution {
    public int[] plusOne(int[] digits) {
        boolean carry = true;
        for(int i = digits.length - 1; i >= 0 && carry; i --) {
            if (digits[i] == 9) {
                digits[i] = 0;
            }
            else {
                digits[i] += 1;
                carry = false;
            }
        }
        if (!carry)
            return digits;
        int[] res = new int[digits.length + 1];
        res[0] = 1;
        return res;
    }
}
